renderdoc/serialise/codecs/vk_cpp_codec/ext_object.h

```diff
@@ -0,0 +1,304 @@
+/******************************************************************************
+* The MIT License (MIT)
+*
+* Copyright (c) 2018 Google LLC
+*
+* Permission is hereby granted, free of charge, to any person obtaining a copy
+* of this software and associated documentation files (the "Software"), to deal
+* in the Software without restriction, including without limitation the rights
+* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+* copies of the Software, and to permit persons to whom the Software is
+* furnished to do so, subject to the following conditions:
+*
+* The above copyright notice and this permission notice shall be included in
+* all copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+* THE SOFTWARE.
+******************************************************************************/
+#pragma once
+
+#include <assert.h>
+#include <math.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <array>
+#include <list>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "common/common.h"
+#include "core/core.h"
+#include "driver/vulkan/vk_common.h"
+#include "driver/vulkan/vk_resources.h"
+#include "serialise/rdcfile.h"
+#include "intervals.h"
+
+namespace vk_cpp_codec
+{
+extern std::unordered_map<uint64_t, std::string> VkImageLayoutStrings;
+
+inline uint32_t as_uint32(uint64_t val)
+{
+  RDCASSERT(val <= UINT32_MAX);
+  return static_cast<uint32_t>(val);
+}
+
+// Extend the SDObject with some API calls to simplify code.
+struct ExtObject : public SDObject
```

Comment:
>>  I'd want removed, I don't want to duplicate all of the container functions for children into the object itself, nor have redundant utility functions.

I am happy to refactor this, but there is a reason some of that code exists. 
Let me go function by function:

PushOne - in contrast to AddChild, this doesn't make a deep copy of the SDObject. It would be good to have an option like that in SDObject. In most cases where this is used, an SDObject is explicitly created and it's members initialized upfront, so doing an additional copy would just be overhead. It would be good to have a more lightweight method. 

RemoveOne - I don't see a method to erase the data.children in SDObject.

Size - I dont see a method to query size of children. There are 224 places where Size() is called in codec now, and changing Size() to data.children.size() everywhere is going to look kind of long. Furthermore, because children size returns size_t, and most of the codec operates with uint64 for simplicity and to keep the code printing to a minimum number of printable types, this will have additional (uint64_t) cast in front of it. 

- Exists, there are only 4 places where this is used so it can go away.

- At - SDObject has the methods I need, but these get chained into At()->At()->At() .. quite a lot. Are you sure you don't want a shorter alias?..

-operator [] - I ended up not using it, so it's gone.

Please let me know what you think of the above. Quick disclaimer, when I added those SDObject didn't have some of those methods which is why I added ExtObject.

I mostly agree with the C++ism and vulkan-ness of the extended class. Let me refactor it a bit

Reply:
I understand that there was good reason to add these helpers and that some of the `SDObject` functions were added later. I appreciate this branch has developed for a long time in isolation and it made sense there to have a separate object class you could iterate on 'internally', but I'm trying to evaluate based on what I think is best to upstream. I'm not sure that all of these functions are something I'd want in `SDObject`.

Regarding `PushOne` / `RemoveOne`, yes there is no direct function in `SDObject`. I meant that I'd prefer direct access to  `data.children` (or `GetChildren()` if you prefer) in these cases. Part of the reason here is that this API can be accessed from python as well and lifetime management becomes difficult, so the default interface I want to have duplicating for that reason. Otherwise you really need to start refcounting `SDObject`s and that's something I want to avoid if at all possible.

Similarly for `Size` I'd prefer just `data.children.size()` or `GetChildren().size()`. I wouldn't mind `NumChildren()` if you prefer for that as it's clear what it means, and doesn't leave a trap for someone thinking `uint64_object->Size()` would return 8.

I'm not sure I understand the case where the cast is needed for `size_t -> uint64_t`, if you have an overloaded function that only accepts `uint64_t`  then `size_t` should be cast to it implicitly regardless of 32-bit/64-bit? I do that a fair bit.

Regarding code compactness mentioned for `Size()` / `At()`, I feel that unless it's really bad then refactoring for shorter code is not the best course of action. As long as it's 'local' and doesn't require jumping around definitions of things I prefer code that leans on the verbose and more clear side if necessary.

E.g. I don't know how long the chains you're talking about are but
```
VkGraphicsPipelineCreateInfo->FindChild("pColorBlendState")->FindChild("pAttachments")
  ->GetChild(0)->FindChild("blendEnable")
```
is not much worse than
```
VkGraphicsPipelineCreateInfo->At("pColorBlendState")->At("pAttachments")->At(0)->At("blendEnable")
```
and feels more intuitive about what's happening to me. In cases where the chain becomes really unmanageable I'd strongly prefer to split it up by declaring intermediate values.

btw one thing I forgot to mention that you might find useful, I added `StructuredSerialiser` for use with indirect draw commands where it's necessary to add extra chunks with the data in the draw command. The usage is like so:

```cpp
SDObject *root_sdobject = ...;

{
  StructuredSerialiser structuriser(root_sdobject, NULL);

  structuriser.Serialise<uint32_t>("drawIndex", 0U);
  structuriser.Serialise("offset", offset);
  structuriser.Serialise("command", VkDrawIndirectCommand());
}
```

where `root_sdobject` is the SDObject that children get added to. The `NULL` parameter is just the chunk lookup function which isn't needed unless you're serialising chunks with this (which is unlikely what you want). The SDObjects for each Serialise call get added as children of `root_sdobject`.

I don't know if it's useful to you, but I thought I'd let you know about it as it's a recent addition.

Reply:
I thought I left a comment here yesterday, but clearly I was wrong :/

>> Regarding PushOne / RemoveOne, yes there is no direct function in SDObject. I meant that I'd prefer direct access to data.children (or GetChildren() if you prefer) in these cases.
Sure, I can do that. There isn't that many places where it's used, maybe 30-40 locations. It's just a visual preference IMHO, because in both cases it calls GetChildren().push_back() or GetChildren().erase() without duplicating SDObject.
Since the number of occurrences is >2 I prefer a more compact code, which also doesn't transcend the code style (i.e. mixing SDObject style with c++ style makes my eyes water). Whether PushOne / RemoveOne are good names is entirely debatable.

>>Part of the reason here is that this API can be accessed from python as well [...]
I agree not all of the manipulations a codec does with the SDObject list should be allowed from python shell. This is why I think codecs should be able to inherit SDObject and extend it with methods, rather than add specialized code to SDObject directly. 

>> [...] and lifetime management becomes difficult, so the default interface I want to have duplicating for that reason. Otherwise you really need to start refcounting SDObjects and that's something I want to avoid if at all possible. 
The code that adds new SDObjects also handles that they are added correctly, i.e. if it's an entirely new SDObject it gets added directly, when it's based of an existing SDObject, that get's duplicated and a dupe gets added. I hope this alleviates your concern re: ref counting.

>> I'm not sure I understand the case where the cast is needed for size_t -> uint64_t
Because of the way size_t is defined differently on platforms, as soon as there is any size_t returned value the code just becomes a constant mess of casting to / from size_t. 
Specifically on MSVC size_t typedef unsigned __int64 size_t;  and the code that I added:
> for(uint64_t i = 0; i < NumChildren(); i++)
>       if(!GetChild(i)->IsInlineable())
compiled fine locally, but of course failed a Windows (!) build in AppVeyor 
> c:\projects\renderdoc\renderdoc\api\replay\structured_data.h(466): error C2220: warning treated as error - no 'object' file
This gets even uglier when you have chains like this:
  SDObject *ptr = root_ptr->GetChild((size_t)indirect_ptr->FindChild("dstBinding")->AsUInt64());
Or when you printf values and occasionally one of the values is a size_t which gets printed as PRIu64 and so on and so forth. I ended up hating size_t with passion and try to hide it as soon as possible so the rest of the code uses stdint.h types. 

>>  I added StructuredSerialiser for use with indirect draw commands 
Yes, I think I saw that, but I think it only happens when a trace is loaded in UI, no?.. I've seen various IndirectDraw[0], [1] commands and such, when the code generator was launched from the UI, which obviously didnt work as expected. If there is a way for me to ignore such objects, I'd like to do that. I thought I could use the "Hidden" flag for it, but I haven't tried it yet.

Reply:
Also, any objection if FindChild() GetChild() have a warning log / assert if the child wasn't found?..

Reply:
I've addressed many of the requests here, but not all. 

I've updated SDObject to include more methods from ExtObject. There are some pieces left:

I'll go over them:

> ExtObject constructors

I removed 2. 1/2 remaining is there to simply call SDObject ctor (since SDObject doesn't have a default constructor) and the second one is to handle the case of creating a ResourceId SDObject. The way the ResourceId structure is implemented it's not possible to call makeSDResourceId because it's not possible to cast an integer value to a resource Id. I am not sure how to work around this and I didn't want to change ResourceId code to make it work. Adding such a fake ResourceId chunk is something that doesn't happen very often, but it happens in places like this:

>wait->PushOne(new ExtObject("aux.semaphore", "VkSemaphore", ACQUIRE_SEMAPHORE_VAR_ID, SDBasic::Resource));

where I am adding a new object that represents a dependency on VkSemaphore that has been acquired from the swapchain.

> void PushOne(ExtObject *o) { data.children.push_back(o); }
> void RemoveOne(ExtObject *o) { data.children.removeOne(o); }
> void RemoveOne(uint64_t i) { data.children.erase(i); }

I can replace these functions with GetChildren.push_back() GetChildren.erase / removeOne if that looks better for you. Renaming them is as easy as search and replace TBH, so it's purely a visual & convenience preference from my side.

> const char *Type()

I added a Type call to SDObject to return vanilla typename, but for VK code gen I need to overload it, in 2 cases. The comment in the code explains why.

> uint64_t &UInt64() { return data.basic.u; }

I left this accessor here, because it didnt feel right to have them in SDObject. These are used in expressions like sdPtr->UInt64() = X; 

>  ExtObject *At(uint64_t i)
 > ExtObject *At(std::string child)

I'd like to have a log messaging when FindChild() failes, and assert happening when i index in GetChild[i] is out of range. Waiting for an OK on that.

>  std::string ValueStr()

Made small fixes to ValueStr() as you recommended.

Reply:
This comment chain is growing out of control, hopefully it's readable via quoted context.

> Since the number of occurrences is >2 I prefer a more compact code, which also doesn't transcend the code style (i.e. mixing SDObject style with c++ style makes my eyes water). Whether PushOne / RemoveOne are good names is entirely debatable.

Naming is a notoriously hard computer science problem. It's not great that the casing is mixed between STL and RenderDoc's public naming scheme but that's not the end of the world and I don't want to add these functions just to avoid that. Please remove them.

> I agree not all of the manipulations a codec does with the SDObject list should be allowed from python shell. This is why I think codecs should be able to inherit SDObject and extend it with methods, rather than add specialized code to SDObject directly.

I don't want to fragment the codebase by having different views of objects in different parts of the code. Codecs can have helper functions as needed, there's no hard reason it must be a derived class with member functions.

> Because of the way size_t is defined differently on platforms, as soon as there is any size_t returned value the code just becomes a constant mess of casting to / from size_t.

I don't think it's because the platforms define it differently, but because linux always compiles 64-bit and doesn't have narrowing conversion warnings included in -Wall.

You can freely cast uint32_t -> size_t regardless of platform or compiler, and you can freely cast size_t -> uint64_t. In this case the warning from visual studio is legitimate, it's warning that on 32-bit your 64-bit index will be truncated.

In general size_t is far from perfect but I use it throughout most of the codebase and it is the 'expected' type for sizes and indices in C++ usually, influenced heavily by the STL. As you've found the need for casts tends to make the choice of one type or another spread outwards so for that reason new code should continue to use size_t for indices or sizes when possible.

E.g. in your code the warning exists because `NumChildren()` is returning uint64_t. Please change it to return `size_t`, then the loop should compile without complaint. FYI also the printf functions in renderdoc support `%zu` if needed for size_t.

If you end up in a situation where you have a uint64_t and you need to use it as an index then you should indeed explicitly cast to size_t. That means most things are still size_t and with the cast you're explicitly stating that you know this uint64_t will be a valid index and won't be potentially too large for size_t. In your example if you prefer you could use `AsUInt32` instead of `AsUInt64` to 'hide' the cast, since you know that `dstBinding` was serialised as `uint32_t` anyway. For example:

```cpp
SDObject *ptr = root_ptr->GetChild(indirect_ptr->FindChild("dstBinding")->AsUInt32());
```

would work fine everywhere.

I would rather have the new code match the rest of the codebase rather than adopting a different style. While I empathise with your opinion, whether you hate size_t or not is irrelevant unless the entire codebase is changed to match an expected style. When adding new code to a project it's more important to be consistent with the additions than to follow personal preference.

> > I added StructuredSerialiser for use with indirect draw commands
> 
> Yes, I think I saw that, but I think it only happens when a trace is loaded in UI, no?.. I've seen various IndirectDraw[0], [1] commands and such, when the code generator was launched from the UI, which obviously didnt work as expected. If there is a way for me to ignore such objects, I'd like to do that. I thought I could use the "Hidden" flag for it, but I haven't tried it yet.

I only mentioned this in case it would make your life easier to use it to generate SDObjects instead of constructing them manually.

And yes unfortunately because the indirect draw chunks rely on actually replaying and reading back the data they aren't present when doing a replay-less conversion, since that might be happening on a machine with no vulkan support at all.

They won't have the hidden flag because that's used for hiding objects from the UI entirely, but your code can skip the chunk type `VulkanChunk::vkCmdIndirectSubCommand` - I could rename it to not contain the `vkCmd` since that is confusing.

> Also, any objection if FindChild() GetChild() have a warning log / assert if the child wasn't found?..

`GetChild` could have a warning for the case where the index is out of bounds, but `FindChild` can be used to determine if a child exists and the naming to me does not suggest that specifying a name that doesn't exist is an error, so I don't want a warning there.

> I removed 2. 1/2 remaining is there to simply call SDObject ctor (since SDObject doesn't have a default constructor) and the second one is to handle the case of creating a ResourceId SDObject. The way the ResourceId structure is implemented it's not possible to call makeSDResourceId because it's not possible to cast an integer value to a resource Id. I am not sure how to work around this and I didn't want to change ResourceId code to make it work. Adding such a fake ResourceId chunk is something that doesn't happen very often, but it happens in places like this:

This is a deliberate choice, ResourceId should be treated as an opaque object and not as an integer, and code should not try to cast an arbitrary integer to a ResourceId. I don't know the context of the example you gave but that's something that should be fixed another way.

> I added a Type call to SDObject to return vanilla typename, but for VK code gen I need to overload it, in 2 cases. The comment in the code explains why.

I don't follow the string case, I guess that's clear to you from the rest of the code but it's not clear from the comment.

The flag bits case I don't understand the problem that it causes, but if it would help to have the original typename in the structured data that shouldn't be hard to do. The definition of `SERIALISE_MEMBER_TYPED` currently just casts the object to the given type and serialises it, meaning the structured data obtains the typename of the casted type.

The macro could add something on the end like `.TypedAs(TypeName<decltype(el.obj)>())` to override the structured data typename with the original. Then you'd still get the flag bits serialised but it would declare the typename `uint32_t`. Because C++ typedefs aren't real types you couldn't get `VkPipelineStageFlags` but perhaps that's enough to fix your case? If you really needed the actual flags type I would be fine with adding manual `.TypedAs("VkPipelineStageFlags")` after each of those flag serialises.

> I left this accessor here, because it didnt feel right to have them in SDObject. These are used in expressions like sdPtr->UInt64() = X;

I'd prefer rather just `sdPtr->data.basic.u = X;` but I'd also be fine with adding setter functions to SDObject.

-----

To be clear, I would like you to completely remove `ExtObject`, the remaining functions should be moved to separate utility helper functions as necessary.

Reply:
>> I don't follow the string case, I guess that's clear to you from the rest of the code but it's not clear from the comment.

All of Vulkan structures use "const char*" for strings parameters. When code generator needs to generate an array of strings it can't use "string" or "std::string" as the type. For example, consider the list of layers that gets passed to vkCreateInstance call. 

    const char*  Layers[2] = {
      /* Layers = */ "VK_LAYER_RENDERDOC_Capture",
      /* Validation Layer = */ "VK_LAYER_LUNARG_standard_validation",
    };

The renderdoc's chunk contains string as the type name, not 'const char*'. When cpp codec parses through the chunk it will return 'const char *' as type.

>> The flag bits case I don't understand the problem that it causes
Consider when cpp codec generates a list of flags for calls like vkQueueSubmit:

    VkPipelineStageFlags pWaitDstStageMask_0[2] = {
      /* pWaitDstStageMask = */ VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT,
      /* pWaitDstStageMask = */ VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT,
    };

Because the serialized type name is stored as VkPipelineStageFlagBits it would produce a list of VkPipelineStageFlagBits, which will not compile correctly. Using uint32_t is a mild improvement but when I looked at vk_serialize.cpp @1457 it appeared that you were already doing some hacky stuff there so it wasn't clear to me how I can get the right type string. Do you mind doing that change so I can rebase onto it?
Also, the traces that were serialized previously, they would still have the old type (VkPipelineStageFlagBits) backed in, wouldn't they?

Everything else is done.

Reply:
OK, I understand the string issue. Then yes this seems like a good candidate for a helper function.

I'll put something together for the flag bits issue. It would be more accurate to reflect the actual type rather than the massaged type to aid stringification.

> Also, the traces that were serialized previously, they would still have the old type (VkPipelineStageFlagBits) backed in, wouldn't they?

No as I mentioned before in another comment chain, there is absolutely no type information stored in captures. The only thing in the captures is the raw data, all type information is applied during interpretation on read so it reflects whatever the latest code does.

Reply:
>I'll put something together for the flag bits issue. It would be more accurate to reflect the actual type rather than the massaged type to aid stringification.
Thanks, that would help.

Reply:
I've pushed a first-draft implementation onto the [vkflags branch](https://github.com/baldurk/renderdoc/tree/vkflags), can you take a look and see if that does what you need? You get things like:

```xml
<array name="pWaitDstStageMask">
	<enum typename="VkPipelineStageFlags" string="VK_PIPELINE_STAGE_ALL_COMMANDS_BIT">65536</enum>
</array>
```

and

```xml
<chunk id="1074" name="vkCmdPipelineBarrier" length="121" threadID="1924" timestamp="4024513" duration="2">
	<ResourceId name="commandBuffer" typename="VkCommandBuffer" width="8" string="ResourceId::136">136</ResourceId>
	<enum name="srcStageMask" typename="VkPipelineStageFlags" string="VK_PIPELINE_STAGE_ALL_COMMANDS_BIT">65536</enum>
	<enum name="destStageMask" typename="VkPipelineStageFlags" string="VK_PIPELINE_STAGE_ALL_COMMANDS_BIT">65536</enum>
	<enum name="dependencyFlags" typename="VkDependencyFlags" string="VkDependencyFlagBits(0)">0</enum>
	...
```

Reply:
Thanks, I browsed through the change and I don't see a change to vkQueueSubmit. The code generator tries to inline code as much as possible, i.e. for your pipeline barrier example above, it would not declare a variable for srcStageMask/destStageMask but generate code like this:

 {
    VkMemoryBarrier* pMemoryBarriers = NULL;
    VkBufferMemoryBarrier* pBufferMemoryBarriers = NULL;
    VkImageMemoryBarrier pImageMemoryBarriers[1] = {
      { /* pImageMemoryBarriers = */
        /* sType = */ VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER,
        /* pNext = */ NULL,
        /* srcAccessMask = */ VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT,
        /* dstAccessMask = */ VK_ACCESS_MEMORY_READ_BIT,
        /* oldLayout = */ VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL,
        /* newLayout = */ VK_IMAGE_LAYOUT_PRESENT_SRC_KHR,
        /* srcQueueFamilyIndex = */ -1,
        /* dstQueueFamilyIndex = */ -1,
        /* image = */ PresentImages_10347[acquired_frame],
        { /* subresourceRange = */
          /* aspectMask = */ VK_IMAGE_ASPECT_COLOR_BIT,
          /* baseMipLevel = */ 0u,
          /* levelCount = */ 4294967295u,
          /* baseArrayLayer = */ 0u,
          /* layerCount = */ 4294967295u,
        },
      },
    };
    vkCmdPipelineBarrier(VkCommandBuffer_815162, VK_PIPELINE_STAGE_ALL_COMMANDS_BIT, VK_PIPELINE_STAGE_ALL_COMMANDS_BIT, VK_DEPENDENCY_BY_REGION_BIT, 0u, pMemoryBarriers, 0u, pBufferMemoryBarriers, 1u, pImageMemoryBarriers);
  }

Because the VK_ACCESS_COLOR_ATTACHMENT_WRITE_BIT is inlined inside the pImageMemoryBarriers[] element it doesn't really matter to the code gen that the type is wrong here.
But for a queue submission for example, I need to generate the array of WaitDstStageMasks like this:
  {
    VkSemaphore pWaitSemaphores_0[1] = {
      /* pWaitSemaphores = */ VkSemaphore_160,
    };
    **VkPipelineStageFlags** pWaitDstStageMask_0[1] = {
      /* pWaitDstStageMask = */ VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT,
    };
    VkCommandBuffer pCommandBuffers_0[1] = {
      /* pCommandBuffers = */ VkCommandBuffer_815162,
    };
    VkSemaphore pSignalSemaphores_0[1] = {
      /* pSignalSemaphores = */ VkSemaphore_160,
    };
    VkSubmitInfo pSubmits_0 = {
      /* sType = */ VK_STRUCTURE_TYPE_SUBMIT_INFO,
      /* pNext = */ NULL,
      /* waitSemaphoreCount = */ 1u,
      /* pWaitSemaphores = */ pWaitSemaphores_0,
      /* pWaitDstStageMask = */ pWaitDstStageMask_0,
      /* commandBufferCount = */ 1u,
      /* pCommandBuffers = */ pCommandBuffers_0,
      /* signalSemaphoreCount = */ 1u,
      /* pSignalSemaphores = */ pSignalSemaphores_0,
    };
    VkSubmitInfo pSubmits[1] = {
      /* pSubmits = */ pSubmits_0,
    };

which causes the compiler to complain about incorrect types. I've specifically looked at the submit call, but I am not sure if it's the only place in vulkan where flags and bits can be confused like this, so the CL is definitely a step in the right direction, but it won't be enough for the code gen to work correctly.

Reply:
Did you try to test with that branch and see if it works? If you were just looking for a change to `vkQueueSubmit` then yes there isn't one, because it's not needed.

I understand the problem, the type information for `VkSubmitInfo::pWaitDstStageMask` comes from the serialisation inside `VkSubmitInfo` which is changed [here in vk_serialise.cpp](https://github.com/baldurk/renderdoc/commit/c4b8b14c334b3de7c535888dc1a2f4b74791e948#diff-9f95601f3f044346b20987bbcde80a02R1607). No change is necessary within the vkQueueSubmit function because all of the type information is set within the serialised structs. The only cases that needed fixups in functions themselves were [cases like vkCmdPipelineBarrier](https://github.com/baldurk/renderdoc/commit/c4b8b14c334b3de7c535888dc1a2f4b74791e948#diff-79720445986ec860f3a279faee4416c9R2490) that was coercing the type of a bitfield function parameter.

Reply:
I tried cherry-picking the specific change but it had a dependency on a previous CL and failed to compile. I'll have to sync my fork to github tot, which can become an effort on it's own. I am happy to remove the flags hack and keep it inside my local fork until I merge to github's tot.

Reply:
If you don't mind doing that then yeh that would be good. I'll merge in the vkflags branch tomorrow and you can pick it up next time you rebase to v1.x latest.

Reply:
I think that would be best.
Also, do you still want it inside codecs\vk_cpp_codec folder? My original thought was to place all vk cpp codec code there, but since most of it will go into the driver directory, and ext_object gets renamed to vk_cpp_codec_helper.h I might as well move it to the codecs directory. Or would you prefer codecs\vk_cpp_codec\vk_cpp_codec_helper.h?.. Please let me know.

Reply:
Yep, I think everything is good now aside from the filename change.

Reply:
Rebased and resolving.